app/master/build_scheduler.py

```diff
@@ -49,7 +49,13 @@ def needs_more_slaves(self):
 
         :rtype: bool
         """
-        return self._num_executors_allocated < self._max_executors and not self._build._unstarted_subjobs.empty()
+        if self._num_executors_allocated >= self._max_executors:
+            return False
+        if self._build._unstarted_subjobs.empty():
+            return False
+        if self._num_executors_allocated >= len(self._build.all_subjobs()):
+            return False
+        return True
```

Comment:
I thought that breaking up the conditionals was much easier to read than all of the 'ands' and 'nots' chained together.